Using a Whole Genome Co-expression Network to Inform the Functional Characterisation of Predicted Genomic Elements from Mycobacterium tuberculosis Transcriptomic Data

Jennifer J. Stiens, Yen Yi Tan, Rosanna Joyce, Kristine Arnvig, Sharon Kendall, Irilenia Nobeli

Abstract

A whole genome co-expression network was created using Mycobacterium tuberculosis transcriptomic data from publicly available RNA-sequencing experiments covering a wide variety of experimental conditions. The network includes expressed regions with no formal annotation, including putative sRNAs and UTRs, along with the protein-coding genes. Non-coding RNA were among the most well-connected members of the module sub-networks, making up more than half of the ‘hub’ genes in modules that include protein-coding genes known to be part of regulatory systems involved in stress response and host adaptation. This dataset provides a valuable resource for investigating the role of non-coding RNA in transcriptomic remodelling. Based on their connections to genes with known functional groupings and correlations with replicated host conditions, predicted non-coding RNA elements can be screened as suitable candidates for further experimental validation.

The summary and overview of scripts relevant for paper is found in Mtb_wgcna_doc.Rmd

Including:

RNA-seq processing and mapping

Transcript prediction

Feature quantification

Normallisation, transformation and batch correction

Network creation with WGCNA

Determining categories and positions of ncRNA transcripts

Further analysis including functional enrichment

https://doi.org/10.5281/zenodo.7319853
